File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-ird-5-4/FxLinkedNotionalAmount.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FxLinkedNotionalAmount : ISerialized
    {
        public FxLinkedNotionalAmount(XmlNode xmlNode)
        {
            XmlNode resetDateNode = xmlNode.SelectSingleNode("resetDate");
            
            if (resetDateNode != null)
            {
                if (resetDateNode.Attributes["href"] != null || resetDateNode.Attributes["id"] != null) 
                {
                    if (resetDateNode.Attributes["id"] != null) 
                    {
                        resetDateIDRef_ = resetDateNode.Attributes["id"].Value;
                        XsdTypeDate ob = new XsdTypeDate(resetDateNode);
                        IDManager.SetID(resetDateIDRef_, ob);
                    }
                    else if (resetDateNode.Attributes["href"] != null)
                    {
                        resetDateIDRef_ = resetDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        resetDate_ = new XsdTypeDate(resetDateNode);
                    }
                }
                else
                {
                    resetDate_ = new XsdTypeDate(resetDateNode);
                }
            }
            
        
            XmlNode adjustedFxSpotFixingDateNode = xmlNode.SelectSingleNode("adjustedFxSpotFixingDate");
            
            if (adjustedFxSpotFixingDateNode != null)
            {
                if (adjustedFxSpotFixingDateNode.Attributes["href"] != null || adjustedFxSpotFixingDateNode.Attributes["id"] != null) 
                {
                    if (adjustedFxSpotFixingDateNode.Attributes["id"] != null) 
                    {
                        adjustedFxSpotFixingDateIDRef_ = adjustedFxSpotFixingDateNode.Attributes["id"].Value;
                        XsdTypeDate ob = new XsdTypeDate(adjustedFxSpotFixingDateNode);
                        IDManager.SetID(adjustedFxSpotFixingDateIDRef_, ob);
                    }
                    else if (adjustedFxSpotFixingDateNode.Attributes["href"] != null)
                    {
                        adjustedFxSpotFixingDateIDRef_ = adjustedFxSpotFixingDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        adjustedFxSpotFixingDate_ = new XsdTypeDate(adjustedFxSpotFixingDateNode);
                    }
                }
                else
                {
                    adjustedFxSpotFixingDate_ = new XsdTypeDate(adjustedFxSpotFixingDateNode);
                }
            }
            
        
            XmlNode observedFxSpotRateNode = xmlNode.SelectSingleNode("observedFxSpotRate");
            
            if (observedFxSpotRateNode != null)
            {
                if (observedFxSpotRateNode.Attributes["href"] != null || observedFxSpotRateNode.Attributes["id"] != null) 
                {
                    if (observedFxSpotRateNode.Attributes["id"] != null) 
                    {
                        observedFxSpotRateIDRef_ = observedFxSpotRateNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(observedFxSpotRateNode);
                        IDManager.SetID(observedFxSpotRateIDRef_, ob);
                    }
                    else if (observedFxSpotRateNode.Attributes["href"] != null)
                    {
                        observedFxSpotRateIDRef_ = observedFxSpotRateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        observedFxSpotRate_ = new XsdTypeDecimal(observedFxSpotRateNode);
                    }
                }
                else
                {
                    observedFxSpotRate_ = new XsdTypeDecimal(observedFxSpotRateNode);
                }
            }
            
        
            XmlNode notionalAmountNode = xmlNode.SelectSingleNode("notionalAmount");
            
            if (notionalAmountNode != null)
            {
                if (notionalAmountNode.Attributes["href"] != null || notionalAmountNode.Attributes["id"] != null) 
                {
                    if (notionalAmountNode.Attributes["id"] != null) 
                    {
                        notionalAmountIDRef_ = notionalAmountNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(notionalAmountNode);
                        IDManager.SetID(notionalAmountIDRef_, ob);
                    }
                    else if (notionalAmountNode.Attributes["href"] != null)
                    {
                        notionalAmountIDRef_ = notionalAmountNode.Attributes["href"].Value;
                    }
                    else
                    {
                        notionalAmount_ = new XsdTypeDecimal(notionalAmountNode);
                    }
                }
                else
                {
                    notionalAmount_ = new XsdTypeDecimal(notionalAmountNode);
                }
            }
            
        
        }
        
    
        #region ResetDate_
        private XsdTypeDate resetDate_;
        public XsdTypeDate ResetDate_
        {
            get
            {
                if (this.resetDate_ != null)
                {
                    return this.resetDate_; 
                }
                else if (this.resetDateIDRef_ != null)
                {
                    resetDate_ = IDManager.getID(resetDateIDRef_) as XsdTypeDate;
                    return this.resetDate_; 
                }
                else
                {
                      return this.resetDate_; 
                }
            }
            set
            {
                if (this.resetDate_ != value)
                {
                    this.resetDate_ = value;
                }
            }
        }
        #endregion
        
        public string resetDateIDRef_ { get; set; }
        #region AdjustedFxSpotFixingDate_
        private XsdTypeDate adjustedFxSpotFixingDate_;
        public XsdTypeDate AdjustedFxSpotFixingDate_
        {
            get
            {
                if (this.adjustedFxSpotFixingDate_ != null)
                {
                    return this.adjustedFxSpotFixingDate_; 
                }
                else if (this.adjustedFxSpotFixingDateIDRef_ != null)
                {
                    adjustedFxSpotFixingDate_ = IDManager.getID(adjustedFxSpotFixingDateIDRef_) as XsdTypeDate;
                    return this.adjustedFxSpotFixingDate_; 
                }
                else
                {
                      return this.adjustedFxSpotFixingDate_; 
                }
            }
            set
            {
                if (this.adjustedFxSpotFixingDate_ != value)
                {
                    this.adjustedFxSpotFixingDate_ = value;
                }
            }
        }
        #endregion
        
        public string adjustedFxSpotFixingDateIDRef_ { get; set; }
        #region ObservedFxSpotRate_
        private XsdTypeDecimal observedFxSpotRate_;
        public XsdTypeDecimal ObservedFxSpotRate_
        {
            get
            {
                if (this.observedFxSpotRate_ != null)
                {
                    return this.observedFxSpotRate_; 
                }
                else if (this.observedFxSpotRateIDRef_ != null)
                {
                    observedFxSpotRate_ = IDManager.getID(observedFxSpotRateIDRef_) as XsdTypeDecimal;
                    return this.observedFxSpotRate_; 
                }
                else
                {
                      return this.observedFxSpotRate_; 
                }
            }
            set
            {
                if (this.observedFxSpotRate_ != value)
                {
                    this.observedFxSpotRate_ = value;
                }
            }
        }
        #endregion
        
        public string observedFxSpotRateIDRef_ { get; set; }
        #region NotionalAmount_
        private XsdTypeDecimal notionalAmount_;
        public XsdTypeDecimal NotionalAmount_
        {
            get
            {
                if (this.notionalAmount_ != null)
                {
                    return this.notionalAmount_; 
                }
                else if (this.notionalAmountIDRef_ != null)
                {
                    notionalAmount_ = IDManager.getID(notionalAmountIDRef_) as XsdTypeDecimal;
                    return this.notionalAmount_; 
                }
                else
                {
                      return this.notionalAmount_; 
                }
            }
            set
            {
                if (this.notionalAmount_ != value)
                {
                    this.notionalAmount_ = value;
                }
            }
        }
        #endregion
        
        public string notionalAmountIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-ird-5-4/FxLinkedNotionalAmount.cpp
// FxLinkedNotionalAmount.cpp 
#include "FxLinkedNotionalAmount.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FxLinkedNotionalAmount::FxLinkedNotionalAmount(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //resetDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* resetDateNode = xmlNode->FirstChildElement("resetDate");

   if(resetDateNode){resetDateIsNull_ = false;}
   else{resetDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- resetDateNode , address : " << resetDateNode << std::endl;
   #endif
   if(resetDateNode)
   {
      if(resetDateNode->Attribute("href") || resetDateNode->Attribute("id"))
      {
          if(resetDateNode->Attribute("id"))
          {
             resetDateIDRef_ = resetDateNode->Attribute("id");
             resetDate_ = boost::shared_ptr<XsdTypeDate>(new XsdTypeDate(resetDateNode));
             resetDate_->setID(resetDateIDRef_);
             IDManager::instance().setID(resetDateIDRef_,resetDate_);
          }
          else if(resetDateNode->Attribute("href")) { resetDateIDRef_ = resetDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { resetDate_ = boost::shared_ptr<XsdTypeDate>(new XsdTypeDate(resetDateNode));}
   }

   //adjustedFxSpotFixingDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* adjustedFxSpotFixingDateNode = xmlNode->FirstChildElement("adjustedFxSpotFixingDate");

   if(adjustedFxSpotFixingDateNode){adjustedFxSpotFixingDateIsNull_ = false;}
   else{adjustedFxSpotFixingDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- adjustedFxSpotFixingDateNode , address : " << adjustedFxSpotFixingDateNode << std::endl;
   #endif
   if(adjustedFxSpotFixingDateNode)
   {
      if(adjustedFxSpotFixingDateNode->Attribute("href") || adjustedFxSpotFixingDateNode->Attribute("id"))
      {
          if(adjustedFxSpotFixingDateNode->Attribute("id"))
          {
             adjustedFxSpotFixingDateIDRef_ = adjustedFxSpotFixingDateNode->Attribute("id");
             adjustedFxSpotFixingDate_ = boost::shared_ptr<XsdTypeDate>(new XsdTypeDate(adjustedFxSpotFixingDateNode));
             adjustedFxSpotFixingDate_->setID(adjustedFxSpotFixingDateIDRef_);
             IDManager::instance().setID(adjustedFxSpotFixingDateIDRef_,adjustedFxSpotFixingDate_);
          }
          else if(adjustedFxSpotFixingDateNode->Attribute("href")) { adjustedFxSpotFixingDateIDRef_ = adjustedFxSpotFixingDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { adjustedFxSpotFixingDate_ = boost::shared_ptr<XsdTypeDate>(new XsdTypeDate(adjustedFxSpotFixingDateNode));}
   }

   //observedFxSpotRateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* observedFxSpotRateNode = xmlNode->FirstChildElement("observedFxSpotRate");

   if(observedFxSpotRateNode){observedFxSpotRateIsNull_ = false;}
   else{observedFxSpotRateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- observedFxSpotRateNode , address : " << observedFxSpotRateNode << std::endl;
   #endif
   if(observedFxSpotRateNode)
   {
      if(observedFxSpotRateNode->Attribute("href") || observedFxSpotRateNode->Attribute("id"))
      {
          if(observedFxSpotRateNode->Attribute("id"))
          {
             observedFxSpotRateIDRef_ = observedFxSpotRateNode->Attribute("id");
             observedFxSpotRate_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(observedFxSpotRateNode));
             observedFxSpotRate_->setID(observedFxSpotRateIDRef_);
             IDManager::instance().setID(observedFxSpotRateIDRef_,observedFxSpotRate_);
          }
          else if(observedFxSpotRateNode->Attribute("href")) { observedFxSpotRateIDRef_ = observedFxSpotRateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { observedFxSpotRate_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(observedFxSpotRateNode));}
   }

   //notionalAmountNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* notionalAmountNode = xmlNode->FirstChildElement("notionalAmount");

   if(notionalAmountNode){notionalAmountIsNull_ = false;}
   else{notionalAmountIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- notionalAmountNode , address : " << notionalAmountNode << std::endl;
   #endif
   if(notionalAmountNode)
   {
      if(notionalAmountNode->Attribute("href") || notionalAmountNode->Attribute("id"))
      {
          if(notionalAmountNode->Attribute("id"))
          {
             notionalAmountIDRef_ = notionalAmountNode->Attribute("id");
             notionalAmount_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(notionalAmountNode));
             notionalAmount_->setID(notionalAmountIDRef_);
             IDManager::instance().setID(notionalAmountIDRef_,notionalAmount_);
          }
          else if(notionalAmountNode->Attribute("href")) { notionalAmountIDRef_ = notionalAmountNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { notionalAmount_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(notionalAmountNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<XsdTypeDate> FxLinkedNotionalAmount::getResetDate()
{
   if(!this->resetDateIsNull_){
        if(resetDateIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDate>(IDManager::instance().getID(resetDateIDRef_));
        }else{
             return this->resetDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDate>();
   }
}
boost::shared_ptr<XsdTypeDate> FxLinkedNotionalAmount::getAdjustedFxSpotFixingDate()
{
   if(!this->adjustedFxSpotFixingDateIsNull_){
        if(adjustedFxSpotFixingDateIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDate>(IDManager::instance().getID(adjustedFxSpotFixingDateIDRef_));
        }else{
             return this->adjustedFxSpotFixingDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDate>();
   }
}
boost::shared_ptr<XsdTypeDecimal> FxLinkedNotionalAmount::getObservedFxSpotRate()
{
   if(!this->observedFxSpotRateIsNull_){
        if(observedFxSpotRateIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(observedFxSpotRateIDRef_));
        }else{
             return this->observedFxSpotRate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
boost::shared_ptr<XsdTypeDecimal> FxLinkedNotionalAmount::getNotionalAmount()
{
   if(!this->notionalAmountIsNull_){
        if(notionalAmountIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(notionalAmountIDRef_));
        }else{
             return this->notionalAmount_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
}

